File: scripts/update-payment-links.py
"""
One-time script to update all Stripe payment links:
 - Enable promotion codes
 - Remove custom fields (delivery date)

Usage:
  STRIPE_SECRET_KEY=sk_live_... python3 scripts/update-payment-links.py
"""

import os
import json
import urllib.request
import urllib.parse
import sys

STRIPE_KEY = os.environ.get('STRIPE_SECRET_KEY', '')
if not STRIPE_KEY:
    print('❌  Set STRIPE_SECRET_KEY env var before running.')
    sys.exit(1)

BASE_URL = 'https://api.stripe.com/v1'

def stripe_req(path, method='GET', params=None):
    url = BASE_URL + path
    data = urllib.parse.urlencode(params).encode() if params else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('Authorization', f'Bearer {STRIPE_KEY}')
    req.add_header('Content-Type', 'application/x-www-form-urlencoded')
    try:
        with urllib.request.urlopen(req) as res:
            return json.loads(res.read())
    except urllib.error.HTTPError as e:
        body = json.loads(e.read())
        return body  # return error body so callers can handle it

def list_all_payment_links():
    links = []
    params = {'limit': '100'}
    while True:
        res = stripe_req('/payment_links?' + urllib.parse.urlencode(params))
        if 'error' in res:
            print(f'❌  Stripe error: {res["error"]["message"]}')
            sys.exit(1)
        links.extend(res['data'])
        if not res.get('has_more'):
            break
        params['starting_after'] = res['data'][-1]['id']
    return links

def update_link(link_id):
    params = {
        'allow_promotion_codes': 'true',
        'custom_fields': '',        # clears all custom fields
    }
    try:
        res = stripe_req(f'/payment_links/{link_id}', method='POST', params=params)
        if 'error' in res:
            print(f'  ❌  Failed: {res["error"]["message"]}')
            return False
        return True
    except Exception as e:
        print(f'  ❌  Exception: {e}')
        return False

if __name__ == '__main__':
    print('🔍  Fetching payment links…')
    links = list_all_payment_links()
    print(f'   Found {len(links)} payment link(s)\n')

    for link in links:
        sys.stdout.write(f'Updating {link["id"]}… ')
        sys.stdout.flush()
        ok = update_link(link['id'])
        print('✓' if ok else '✗')

    print('\n✅  Done.')

File: scripts/update-payment-links.mjs
/**
 * One-time script to update all Stripe payment links:
 *  - Enable promotion codes
 *  - Remove custom fields (delivery date)
 *
 * Usage:
 *   STRIPE_SECRET_KEY=sk_live_... node scripts/update-payment-links.mjs
 */

const STRIPE_KEY = process.env.STRIPE_SECRET_KEY;
if (!STRIPE_KEY) {
  console.error('❌  Set STRIPE_SECRET_KEY env var before running.');
  process.exit(1);
}

async function stripeReq(path, method = 'GET', params) {
  const res = await fetch(`https://api.stripe.com/v1${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${STRIPE_KEY}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: params?.toString(),
  });
  return res.json();
}

// Fetch all payment links (up to 100)
async function listAllPaymentLinks() {
  const links = [];
  let hasMore = true;
  let startingAfter = null;

  while (hasMore) {
    const params = new URLSearchParams({ limit: '100' });
    if (startingAfter) params.set('starting_after', startingAfter);

    const res = await stripeReq(`/payment_links?${params.toString()}`);
    if (res.error) {
      console.error('❌  Error listing payment links:', res.error.message);
      process.exit(1);
    }

    links.push(...res.data);
    hasMore = res.has_more;
    if (hasMore) startingAfter = res.data[res.data.length - 1].id;
  }

  return links;
}

async function updateLink(id) {
  const params = new URLSearchParams();
  params.set('allow_promotion_codes', 'true');
  // Clear all custom fields by passing an empty array
  params.set('custom_fields', '');

  const result = await stripeReq(`/payment_links/${id}`, 'POST', params);
  if (result.error) {
    console.error(`  ❌  Failed: ${result.error.message}`);
    return false;
  }
  return true;
}

(async () => {
  console.log('🔍  Fetching payment links…');
  const links = await listAllPaymentLinks();
  console.log(`   Found ${links.length} payment link(s)\n`);

  for (const link of links) {
    const label = link.metadata?.label || link.id;
    process.stdout.write(`Updating ${link.id}… `);
    const ok = await updateLink(link.id);
    console.log(ok ? '✓' : '✗');
  }

  console.log('\n✅  Done. All payment links updated.');
})();
